torch/autograd/gradcheck.py

```diff
@@ -368,81 +541,35 @@ def fail_test(msg):
-                return _as_tuple(func(*input))[i]
-            numerical = get_numerical_jacobian(fn, tupled_inputs, eps=eps)
-            for n in numerical:
-                if torch.ne(n, 0).sum() > 0:
-                    return fail_test('Numerical gradient for function expected to be zero')
-        return True
+    outputs = _differentiable_outputs(func_out)
 
-    for i, o in enumerate(output):
-        if not o.requires_grad:
-            continue
+    check_outputs(outputs)
+
+    if not outputs:
+        return check_no_differentiable_outputs(fail_test, func, tupled_inputs, func_out, eps)
 
+    for i, o in enumerate(outputs):
         def fn(input):
             return _as_tuple(func(*input))[i]
 
-        analytical, reentrant, correct_grad_sizes, correct_grad_types = get_analytical_jacobian(tupled_inputs,
-                                                                                                o,
-                                                                                                nondet_tol=nondet_tol)
+        analytical, failed = check_analytical_jacobian(tupled_inputs, o, nondet_tol, 1.0, check_batched_grad,
+                                                       check_grad_dtypes, raise_exception)
+        if failed:
+            return False
         numerical = get_numerical_jacobian(fn, tupled_inputs, eps=eps)
 
-        out_is_complex = o.is_complex()
-
-        if out_is_complex:
-            # analytical vjp with grad_out = 1.0j
-            analytical_with_imag_grad_out, reentrant_with_imag_grad_out, \
-                correct_grad_sizes_with_imag_grad_out, correct_grad_types_with_imag_grad_out \
-                = get_analytical_jacobian(tupled_inputs, o, nondet_tol=nondet_tol, grad_out=1j)
-            numerical_with_imag_grad_out = get_numerical_jacobian(fn, tupled_inputs, eps=eps, grad_out=1j)
-
-        if not correct_grad_types and check_grad_dtypes:
-            return fail_test('Gradient has dtype mismatch')
-
-        if out_is_complex and not correct_grad_types_with_imag_grad_out and check_grad_dtypes:
-            return fail_test('Gradient (calculated using complex valued grad output) has dtype mismatch')
-
-        if not correct_grad_sizes:
-            return fail_test('Analytical gradient has incorrect size')
-
-        if out_is_complex and not correct_grad_sizes_with_imag_grad_out:
-            return fail_test('Analytical gradient (calculated using complex valued grad output) has incorrect size')
+        if o.is_complex():
+            analytical_imag_grad_out, failed = check_analytical_jacobian(tupled_inputs, o, nondet_tol, 1j,
+                                                                         check_batched_grad, check_grad_dtypes,
+                                                                         raise_exception)
+            if failed:
+                return False
+            numerical_imag_grad_out = get_numerical_jacobian(fn, tupled_inputs, eps=eps, grad_out=1j)
```

Comment:
Does `checkIfNumericalAnalyticAreClose` get used at all in the new code? If not, this probably means we don't have any gradcheck tests lol.

Reply:
Its no longer used in the next PR in the stack, but it is still used in this one :468 for example. Indeed there is no gradcheck test catching the fact that we don't actually return the result of `checkIfNumericalAnalyticAreClose` (similarly for `test_batched_grad`), but I choose to add the return and changes nearby to that PR so this one could ideally have no behavior change.

Reply:
Oh, *facepalm, the way the blame shows up I read it as like 468 gets called from inside `checkIfNumericAnalyticAreClose`. Thanks for clarifying this, that makes sense